src/Stack/Nix.hs

```diff
@@ -0,0 +1,180 @@
+  do config <- asks getConfig
+     inShell <- getInShell
+     isReExec <- asks getReExec
+     if | inShell && not isReExec ->
+            throwM OnlyOnHostException
+        | inShell ->
+            liftIO (do inner
+                       exitSuccess)
+        | not (nixEnable (configNix config)) ->
+            do liftIO inner
+               liftIO exitSuccess
+        | otherwise ->
+            do runShellAndExit
+                 getCmdArgs
+
+runShellAndExit :: M env m
+                => m (String, [String])
+                -> m ()
+runShellAndExit getCmdArgs = do
+     config <- asks getConfig
+     envOverride <- getEnvOverride (configPlatform config)
+     (cmnd,args) <- getCmdArgs
+     isStdoutTerminal <- asks getTerminal
+     (isStdinTerminal,isStderrTerminal) <-
+       liftIO ((,) <$> hIsTerminalDevice stdin
+                   <*> hIsTerminalDevice stderr)
+     let mshellFile = nixInitFile (configNix config)
+         pkgsInConfig = nixPackages (configNix config)
+     if not (null pkgsInConfig) && isJust mshellFile then
+       throwM NixCannotUseShellFileAndPackagesException
+       else return ()
+     let isTerm = isStdinTerminal && isStdoutTerminal && isStderrTerminal
+         nixopts = case mshellFile of
+           Just filePath -> [filePath]
+           Nothing -> ["-E", intercalate " " $ concat
+                              [["with (import <nixpkgs> {});"
+                               ,"runCommand \"myEnv\" {"
+                               ,"buildInputs=lib.optional stdenv.isLinux glibcLocales ++ lib.optional stdenv.isDarwin darwin.cf-private ++ ["],pkgsInConfig,["];"
+                               ,"shellHook=''"
+                               ,   "STACK_IN_NIX_EXTRA_ARGS='"]
+                               ,      (map (\p -> concat ["--extra-lib-dirs=", "${"++p++"}/lib"
+                                                         ," --extra-include-dirs=", "${"++p++"}/include"])
+                                           pkgsInConfig), ["' ;"
+                               ,"'';"
+                               ,"} \"\""]]]
+                    -- glibcLocales is necessary on Linux to avoid warnings about GHC being incapable to set the locale.
+         fullArgs = concat [ -- ["--pure"],
+                            map T.unpack (nixShellOptions (configNix config))
+                           ,nixopts
+                           ,["--command", ("export " ++ inShellEnvVar ++ "=1 ; ")
+                                          ++ intercalate " " (cmnd:args)
+                                          ++ " $STACK_IN_NIX_EXTRA_ARGS"]
+                           ]
+     $logDebug $ T.pack $
+         "Using a nix-shell environment " ++ (case mshellFile of
+            Just filePath -> "from file: " ++ filePath
+            Nothing -> "with nix packages: " ++ (intercalate ", " pkgsInConfig))
+     e <- try (callProcess'
+                 (if isTerm then id else \cp -> cp { delegate_ctlc = False })
```

Comment:
I doubt the extra logic to prevent delegating Ctrl-C is necessary in this case.  For Docker, it's there because of some specific weird stuff that `docker run` and `docker start` do with container terminals, but this nix-shell looks like a regular process that uses the existing `std*` file handles and TTY normally.